### PR TITLE
Add experimental capture checking

### PR DESCRIPTION
**Pull request by odersky**
This PR merges branch cc-experiment into main. The capture checking language extensions are all under a -Ycc flag. At the present stage, they should be considered highly experimental and unstable.

Nevertheless, I think it makes sense to have capture checking as part of the main branch. It will encourage more experiments and will make it easier to incorporate improvements in a stable manner. Since we will have several new people working on capture checking starting next month, it's a good time to merge now.

### CODE REVIEW

**Comment by odersky** [comment]
@Linyxus I put you down as a reviewer since it might be a good way to get started to work on further improvements, which would come in follow-on PRs. Don't worry if there are things which are unclear for now. But if you spot mistakes and possible improvements please report them in the review. The review is best done by comparing the diffs for all commits together. With 97 commits, it would be hard to review them one-by-one, in particular since later commits often change what was added in earlier commits.

If you can only do part of it, that's fine as well. I realize it's a big chunk of work to review more than 8000 lines of code. If others want to chip in, please go ahead!

File: compiler/src/dotty/tools/dotc/cc/CaptureOps.scala [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+   *  intersections, and type proxies other than abstract types.
+   */
+  def boxedCaptureSet(using Context): CaptureSet =
+    def getBoxed(tp: Type): CaptureSet = tp match
+      case tp @ CapturingType(parent, refs) =>
+        val pcs = getBoxed(parent)
+        if tp.isBoxed then refs ++ pcs else pcs
+      case tp: TypeRef if tp.symbol.isAbstractType => CaptureSet.empty
+      case tp: TypeProxy => getBoxed(tp.superType)
+      case tp: AndType => getBoxed(tp.tp1) ++ getBoxed(tp.tp2)
```
Comment (Linyxus) [inline compiler/src/dotty/tools/dotc/cc/CaptureOps.scala:91]: I am not sure whether this is correct. Should we use `++` for unions and `**` for intersections?

**Comment by odersky** [comment]
Going to merge now, so that it becomes easier to deal with new issues.

File: tests/semanticdb/metac.expect [hunk 2]
```diff
@@ -359,7 +359,6 @@ Text => empty
 Language => Scala
 Symbols => 14 entries
 Occurrences => 30 entries
-Synthetics => 2 entries
```
Comment (tgodzik) [inline tests/semanticdb/metac.expect:362]: @odersky Why were all those entries removed? Could we revert the changes?
